scripts/maps_score.py

```diff
@@ -0,0 +1,198 @@
+
+    def get_criteria(i):
+        if singleton:
+            return hl.int(ht.freq[i].AC == 1)
+        elif impose_high_af_cutoff:
+            return hl.int((ht.freq[i].AC > 0) & (ht.freq[i].AF <= 0.001))
+        else:
+            return hl.int(ht.freq[i].AC > 0)
+
+    return hl.agg.array_sum(hl.map(get_criteria, sorted_indices))
+
+
+def count_variants(
+    ht: hl.Table,
+    count_singletons: bool = False,
+    count_downsamplings=(),
+    additional_grouping=(),
+    partition_hint: int = 100,
+    omit_methylation: bool = False,
+    return_type_only: bool = False,
+    force_grouping: bool = False,
+    singleton_expression: hl.expr.BooleanExpression = None,
+    impose_high_af_cutoff_here: bool = False,
+):
+    """
+    Count variants by context, ref, alt, methylation_level
+    """
+
+    grouping = hl.struct(context=ht.context, ref=ht.ref, alt=ht.alt)
+    if not omit_methylation:
+        grouping = grouping.annotate(methylation_level=ht.methylation_level)
+    for group in additional_grouping:
+        grouping = grouping.annotate(**{group: ht[group]})
+
+    if count_singletons:
+        # singleton = hl.any(
+        #    lambda f: (f.meta.size() == 1)
+        #    & (f.meta.get('group') == 'adj')
+        #    & (f.AC[1] == 1),
+        #    ht.freq,
+        # )
+        if singleton_expression is None:
+            singleton_expression = ht.freq[0].AC == 1
+
+    if count_downsamplings or force_grouping:
+        # Slower, but more flexible (allows for downsampling agg's)
+        output = {
+            'variant_count': hl.agg.count_where(ht.freq[0].AF <= 0.001)
+            if impose_high_af_cutoff_here
+            else hl.agg.count()
+        }
+        for pop in count_downsamplings:
+            output[f'downsampling_counts_{pop}'] = downsampling_counts_expr(
+                ht, pop, impose_high_af_cutoff=impose_high_af_cutoff_here
+            )
+        if count_singletons:
+            output['singleton_count'] = hl.agg.count_where(singleton_expression)
+            for pop in count_downsamplings:
+                output[
+                    f'singleton_downsampling_counts_{pop}'
+                ] = downsampling_counts_expr(ht, pop, singleton=True)
+        return (
+            ht.group_by(**grouping).partition_hint(partition_hint).aggregate(**output)
+        )
+    else:
+        agg = {'variant_count': hl.agg.counter(grouping)}
+        if count_singletons:
+            agg['singleton_count'] = hl.agg.counter(
+                hl.agg.filter(singleton_expression, grouping)
+            )
+
+        if return_type_only:
+            return agg['variant_count'].dtype
+        else:
+            return ht.aggregate(hl.struct(**agg))
+
+
+def maps(
+    ht: hl.Table,
+    mutation_ht: hl.Table,
+    additional_grouping: List[str] = [],
+    singleton_expression: hl.expr.BooleanExpression = None,
+    skip_worst_csq: bool = False,
+) -> hl.Table:
+    if not skip_worst_csq:
```

Comment:
A docstring would be useful here as well. Additionally, I'd comment each step in the function body, unless it follows some standard flow of calculating this MAPS score, in which case I'd put a link to the corresponding publication.

Reply:
I added the github file & line number of this function. Hope that'll be enough